File: OSCommon/org/optimizationservices/oscommon/communicationagent/OSRegistryAgent.cs
using System;
using System.Diagnostics;
using System.Xml.Serialization;
using System.Web.Services.Protocols;
using System.ComponentModel;
using System.Web.Services;

using org.optimizationservices.oscommon.communicationinterface;

namespace org.optimizationservices.oscommon.communicationagent{
	/// <summary>
	/// The <c>OSRegistryAgent</c> class implements all the methods in 
	/// <c>OSdL</c> interface. Thus it is OS-type registry agent. 
	/// It implements all the methods as specified in the
	/// OS Framework, specifically the OSdL Language (OSdL).
	/// The <c>OSRegistryAgent</c> class is a library class that 
	/// contains methods to help registry agents communicate with an
	/// OS Registry as specified by the OS framework, in particular it allows clients
	/// to find and register OS Services and validate OSxL arguments. It hides all 
	/// the SOAP protocol related technical details from an optimization user.
	/// @author Jun Ma	
	/// @version 1.0, 09/01/2005
	/// @since OS 1.0
	/// @copyright (c) 2005
	/// </summary>
	[System.Diagnostics.DebuggerStepThroughAttribute()]
	[System.ComponentModel.DesignerCategoryAttribute("code")]
	[System.Web.Services.WebServiceBindingAttribute(
		 Name="OSRegistryServiceSoapBinding", 
		 Namespace="http://www.optimizationservices.org")]
	public class OSRegistryAgent : System.Web.Services.Protocols.SoapHttpClientProtocol, OSdL {
		public string registryAddress{
			get{
				return this.Url;
			}
			set{
				this.Url = value;
			}
		}//registryAddress

		/// <summary>
		/// default constructor
		/// </summary>
		public OSRegistryAgent(){
		}//constructor

		/// <summary>
		/// constructor
		/// </summary>
		/// <param name="registryAddress">registryAddress holds the address (uri/url) of the OS registry.</param>
		public OSRegistryAgent(string registryAddress){
			this.Url = registryAddress;
		}//constructor

		/// <summary>
		/// Find a service at an OS Registry service whose input query is given by a string following 
		/// the OSqL schema, along with an option that follows the OSoL schema,
		/// and returns the uri/url result in a string that follows OSuL schema. 
		/// </summary>
		/// <param name="osql">holds the instance in a string which format follows OSqL schema</param>
		/// <param name="osol">holds the option in a string which format follows OSoL schema</param>
		/// <returns>osul, the uri/url result in a string which format follows the OSuL schema</returns>
		[System.Web.Services.Protocols.SoapRpcMethodAttribute("OSRegistryService#find", 
			 RequestNamespace="http://www.optimizationservices.org", 
			 ResponseNamespace="http://www.optimizationservices.org")]
		[return: System.Xml.Serialization.SoapElementAttribute("osul")]
		public string find(string osql, string osol) {
			object[] results = this.Invoke("find", new object[] {osql, osol});
			return ((string)(results[0]));
		}//find

		/// <summary>
		/// Register a service at an OS Registry service whose input is given by a string following 
		/// the OSeL schema, along with an option that follows the OSoL schema,
		/// and returns the registration process output in a string that follows OSpL schema. 
		/// </summary>
		/// <param name="osel">holds the service entity information in a string which format follows OSeL schema</param>
		/// <param name="osol">holds the option in a string which format follows OSoL schema</param>
		/// <returns>ospl, the registration process output in a string which format follows the OSpL schema</returns>
		[System.Web.Services.Protocols.SoapRpcMethodAttribute("OSRegistryService#register", 
			 RequestNamespace="http://www.optimizationservices.org", 
			 ResponseNamespace="http://www.optimizationservices.org")]
		[return: System.Xml.Serialization.SoapElementAttribute("ospl")]
		public string register(string osel, string osol) {
			object[] results = this.Invoke("register", new object[] {osel, osol});
			return ((string)(results[0]));
		}//register
        
		/// <summary>
		/// Validates an OSxL string at an OS Registry service whose input is any OSxL string following 
		/// the corresponding OSxL schema, along with an option that follows the OSoL schema,
		/// and returns the validation process output in a string that follows OSpL schema. 
		/// </summary>
		/// <param name="osxl">holds the OSxL string which format follows the corresponding schema</param>
		/// <param name="osol">holds the option in a string which format follows OSoL schema</param>
		/// <returns>ospl, the validation process output in a string which format follows the OSpL schema</returns>
		[System.Web.Services.Protocols.SoapRpcMethodAttribute("OSRegistryService#validate", 
			 RequestNamespace="http://www.optimizationservices.org", 
			 ResponseNamespace="http://www.optimizationservices.org")]
		[return: System.Xml.Serialization.SoapElementAttribute("ospl")]
		public string validate(string osxl, string osol) {
			object[] results = this.Invoke("validate", new object[] {osxl, osol});
			return ((string)(results[0]));
		}//validate

		/// <summary>
		/// Set (push) and/or get (pull) runtime dynamic process information to and/or from the service.  		
		/// </summary>
		/// <param name="osplInput"> the input process information in a string which format follows OSpL schema.</param>
		/// <param name="osol">holds the option in a string which format follows OSoL schema</param>
		/// <returns>the output process information in a string which format follows the OSpL schema.</returns>
		[System.Web.Services.Protocols.SoapRpcMethodAttribute("OSRegistryService#knock", 
			 RequestNamespace="http://www.optimizationservices.org", 
			 ResponseNamespace="http://www.optimizationservices.org")]
		[return: System.Xml.Serialization.SoapElementAttribute("osplOutput")]
		public string knock(string osplInput, string osol) {
			object[] results = this.Invoke("knock", new object[] {osplInput, osol});
			return ((string)(results[0]));
		}//knock


	}//class OSRegistryAgent
}//namespace

File: OSjava/OSCommon/src/org/optimizationservices/oscommon/communicationagent/OSRegistryAgent.java
/**
 * @(#)OSRegistryAgent 1.0 03/14/2004
 *
 * Copyright (c) 2004
 */
package org.optimizationservices.oscommon.communicationagent;

import org.optimizationservices.oscommon.communicationinterface.OSdL;
import org.optimizationservices.oscommon.util.OSParameter;
import org.optimizationservices.oscommon.util.WSUtil;


/**
 *
 * <P>The <code>OSRegistryAgent</code> class implements all the <code>OSdL</code> interface.
 * Thus it is OS-type registry. It implements all the methods as specified in the
 * Optimization Services (OS) Framework, specifically the Optimization Services discover
 * Language (OSdL).</p> 
 * 
 * <P>The <code>OSRegistryAgent</code> class is a library class that contains methods
 * to help registry agents communicate with Optimization Services (OS) registries as specified
 * by the Optimization Services framework. It hides all the SOAP protocol related technical details
 * from an optimization user.</p>
 *
 * </p>
 *
 * @author Robert Fourer, Jun Ma, Kipp Martin
 * @version 1.0, 03/14/2004
 * @see org.optimizationservices.oscommon.communicationinterface.OSdL
 * @since OS 1.0
 */
public class OSRegistryAgent implements OSdL{
	/**
	 * registryAddress holds the address (uri/url) of the Optimization Services (OS) registry.
	 */
	public String registryAddress = "";
	
	/**
	 * Constructor. 
	 *
	 */
	public OSRegistryAgent(){
		
	}//constructor
	
	/**
	 * Constructor.
	 * 
	 * @param registryAddress holds the address (uri/url) of the Optimization Services (OS) registry.
	 */
	public OSRegistryAgent(String registryAddress){
		this.registryAddress = registryAddress;
	}//constructor
	
	/**
	 * Discover an Optimization Service whose instance is given by a string following 
	 * the Optimization Services query Language (OSqL) schema and returns the location information
	 * in a string that follows the Optimization Services URI, or Universal Resource Indentifier 
	 * Language, (OSuL) schema.
	 * 
	 * </P>
	 *
	 * @param OSqL holds the Optimization Service query in a string which format follows the 
	 * Optimization Services query Language (OSqL) schema.  
	 * @param osol holds the optimization option in a string which format follows the 
	 * Optimization Services option Language (OSoL) schema.    
	 * @return the location information in a string that follows the Optimization Services URI, 
	 * or Universal Resource Indentifier Language, (OSuL) schema.
	 */
	public String find(String osql, String osol){
		String sMethod = "find";
		String[] msInputs = new String[]{osql, osol};
		String[] msInputNames = new String[]{"osql", "osol"};
		String sReturnName = "osul"; 
		String sSoapAction = "OSRegistryService#find";
		String sOSuL = WSUtil.invokeOptimizationServices(registryAddress, sMethod, msInputs, msInputNames, sReturnName, sSoapAction);
		return sOSuL;
	}//find
	
	/**
	 * register an Optimization Service whose instance is given by a string following 
	 * the Optimization Services entity Language (OSeL) schema and returns a string that contains 
	 * an XSLT style sheet transformation so that it can be used for standard web pubblication of the 
	 * Optimization Service that just joined. The XSLT is espcially used to publish the OSeL. 
	 * 
	 * </P>
	 * 
	 * @param osel holds the Optimization Service entity description in a string which format follows the 
	 * Optimization Services entity Language (OSeL) schema.  
	 * @param osol holds the optimization option in a string which format follows the 
	 * Optimization Services option Language (OSoL) schema.    
	 * @return the register process output information in a string which format follows the OSpL schema.
	 */
	public String register(String osel, String osol){
		String sMethod = "register";
		String[] msInputs = new String[]{osel, osol};
		String[] msInputNames = new String[]{"osel", "osol"};
		String sReturnName = "ospl"; 
		String sSoapAction = "OSRegistryService#register";
		String sOSpL = WSUtil.invokeOptimizationServices(registryAddress, sMethod, msInputs, msInputNames, sReturnName, sSoapAction);
		return sOSpL;
	}//register

	/**
	 * Validate an osxl string according to the corresponding OSxL schema.
	 *
	 * </P>
	 *
	 * @param osxl holds the xml string to be validated. 
	 * @param osol holds the optimization option in a string which format follows the 
	 * Optimization Services option Language (OSoL) schema.    
	 * @return the validate process output information in a string which format follows the OSpL schema.
	 */
	public String validate(String osxl, String osol){
		String sMethod = "validate";
		String[] msInputs = new String[]{osxl, osol};
		String[] msInputNames = new String[]{"osxl", "osol"};
		String sReturnName = "ospl"; 
		String sSoapAction = "OSRegistryService#validate";
		String sOSpL = WSUtil.invokeOptimizationServices(registryAddress, sMethod, msInputs, msInputNames, sReturnName, sSoapAction);
		return sOSpL;
	}//validate

	/**
	 * This method is called by one optimization service on another optimization service
	 * to provide runtime dynamic process information (push mechanism).
	 *
	 * </P>
	 * @param osplInput holds the input process information a string which format follows the Optimization
	 * Services process Language (OSpL) schema.
	 * @param osol holds the optimization option in a string which format follows the 
	 * Optimization Services option Language (OSoL) schema.    
	 * @return osplOutput, the ouput process information in a string which format follows the OSpL schema.
	 */
	public String knock(String osplInput, String osol){
		String sMethod = "knock";		
		String[] msInputs = new String[]{osplInput, (osol==null)?"":osol};
		String[] msInputNames = new String[]{"osplInput", "osol"};
		String sReturnName = "osplOutput"; 
		String sSoapAction = "OSRegistryService#knock";
		String sOSpLOutput = WSUtil.invokeOptimizationServices(registryAddress, sMethod, msInputs, msInputNames, sReturnName, sSoapAction);
		return sOSpLOutput;
	}//knock

	/**
	 * main for test purposes.
	 *
	 * </p>
	 *
	 * @param argv command line arguments.
	 */
	public static void main(String [] args) {
		String sOSqL = "<OSqL>blah blah registry blah blah</OSqL>";
		OSRegistryAgent osRegistryAgent = new OSRegistryAgent();
		osRegistryAgent.registryAddress = OSParameter.OS_REGISTRY_SITE;
		String sOSuL = osRegistryAgent.find(sOSqL, null);
		System.out.println(sOSuL);
	}//main
	
}//class OSRegistryAgent
